bee-crypto/src/ternary/sponge/curlp.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use crate::ternary::{sponge::Sponge, HASH_LENGTH};
+
+use bee_ternary::{Btrit, TritBuf, Trits};
+
+use std::{
+    convert::Infallible,
+    ops::{Deref, DerefMut},
+};
+
+const STATE_LENGTH: usize = HASH_LENGTH * 3;
+const HALF_STATE_LENGTH: usize = STATE_LENGTH / 2;
+
+const TRUTH_TABLE: [[Btrit; 3]; 3] = [
+    [Btrit::PlusOne, Btrit::Zero, Btrit::NegOne],
+    [Btrit::PlusOne, Btrit::NegOne, Btrit::Zero],
+    [Btrit::NegOne, Btrit::PlusOne, Btrit::Zero],
+];
+
+/// Available round numbers for CurlP.
+#[derive(Copy, Clone)]
+pub enum CurlPRounds {
+    /// 27 rounds.
+    Rounds27 = 27,
+    /// 81 rounds.
+    Rounds81 = 81,
+}
+
+/// State of the ternary cryptographic function `CurlP`.
+pub struct CurlP {
+    /// The number of rounds of hashing to apply to the state.
+    rounds: CurlPRounds,
+    /// The internal state.
+    state: TritBuf,
+    /// Workspace for performing transformations.
+    work_state: TritBuf,
+}
+
+impl CurlP {
+    /// Create a new `CurlP` sponge with `rounds` of iterations.
+    pub fn new(rounds: CurlPRounds) -> Self {
+        Self {
+            rounds,
+            state: TritBuf::zeros(STATE_LENGTH),
+            work_state: TritBuf::zeros(STATE_LENGTH),
+        }
+    }
+
+    /// Transforms the internal state of the `CurlP` sponge after the input was copied into the internal state.
+    ///
+    /// The essence of this transformation is the application of a substitution box to the internal state, which happens
+    /// `rounds` number of times.
+    fn transform(&mut self) {
+        #[inline]
+        fn truth_table_get(state: &Trits, p: usize, q: usize) -> Btrit {
+            // Safe to unwrap since indexes come from iteration on the state.
+            TRUTH_TABLE[(state.get(q).unwrap() as i8 + 1) as usize][(state.get(p).unwrap() as i8 + 1) as usize]
+        }
```

Comment:
Can we make this faster with using `get_unchecked`, or is the compiler smart enough to recognize we're always within bounds?

Reply:
'get_unchecked' is 'unsafe' though, do you think it's alright ?

Reply:
why not just `state[q]`?

Reply:
well, `unsafe` just means we know what we're doing :stuck_out_tongue: If we know the passed in indexes can never go out of bounds, then we can use that to our advantage. Maybe @zesterer wants to comment on this?

Reply:
Yeah actually `state[q]` is fine :)

Reply:
just in case, `state[q]` compiles to basically the same to `state.get(q).unwrap()` but it has a nicer panic message, it still checks that the index is out of bounds using an assert. Using `state.get_unchecked(q)` would indeed be faster but then I'd tag the `truth_table_get` function as unsafe just to be sure that everyone is sure that they aren't causing UB by mistake... having a silent OOB just to squeeze a bit of performance is a tricky decision

Reply:
This function is nested in another function anyway, nobody is going to call it except the super function. And we know for sure indexed are fine since we're iterating over the state.